PyTorch AMD Build Scripts

Jorghi12 opened the pull request:
The PyTorch script files necessary for building on AMD GPUs.

Jorghi12 committed:
PyTorch AMD Build Script.

Jorghi12 committed:
Python invocation for hipify

Jorghi12 committed:
Adding individual hip fles.

ezyang commented:
Hi @Jorghi12, does this mean we are ready to setup AMD builders on CI?

Also, directory structure bikeshedding: can we put this inside `tools`, e.g., as `tools/amd`?

ezyang commented on amd_build/build_pytorch_amd.py:
```diff
@@ -0,0 +1,47 @@
+import subprocess
+import os
+
+cwd = os.getcwd()
```
It would be more robust to use `__file__` to determine where the script lives; otherwise this script is sensitive to the particular directory you run it in.

ezyang commented on amd_build/build_pytorch_amd.py:
```diff
@@ -0,0 +1,47 @@
+
+cwd = os.getcwd()
+proj_dir = os.path.dirname(cwd)
+out_dir = os.path.join(os.path.dirname(proj_dir), "pytorch_amd")
```
I don't see updates to `.gitignore`. Is the copied source tree ignored appropriately by git?

ezyang commented on amd_build/build_pytorch_amd.py:
```diff
@@ -0,0 +1,47 @@
+shutil.copy(os.path.join(cwd, "hip_files/THC/generic/THCTensorRandom.cu.hip"), os.path.join(out_dir, "aten/src/THC/generic/THCTensorRandom.cu"))
+
+# Move to avoid HCC bug.
+shutil.move(os.path.join(out_dir, "aten/src/ATen/native/cudnn/Conv.cpp"), os.path.join(out_dir, "aten/src/ATen/native/cudnn/ConvCuDNN.cpp"))
```
I'm OK with renaming these files in the native directory ourselves and then removing the special case here.

ezyang commented on amd_build/build_pytorch_amd.py:
```diff
@@ -0,0 +1,47 @@
+
+# Execute the Hipify Script.
+subprocess.Popen(
+    ["/opt/rocm/bin/hipify-python.py",
```
Most people aren't going to have this script. How do I get it?

Jorghi12 replied:
It's going to be on the HIP Codebase here: https://github.com/ROCm-Developer-Tools/pyHIPIFY, obtained from installing HIP/ROCm packages.

ezyang replied:
OK, don't tell me, put it in the code so other people can find it :)

csuji replied:
@Jorghi12 Hi, I am trying to build your pytorch sending_pr branch for AMD but it seams the current pyHIPIFY is not up to date, option --include-dirs missing and I get the following error:
```
Total number of replaced kernel launches: 1433
Traceback (most recent call last):
  File "/opt/rocm/bin/hipify-python.py", line 845, in <module>
    main()
  File "/opt/rocm/bin/hipify-python.py", line 809, in main
    txt = disable_function(txt, func, 1)
  File "/opt/rocm/bin/hipify-python.py", line 329, in disable_function
    "return_type": the_match.group(2).strip(),
AttributeError: 'NoneType' object has no attribute 'group'
```
and is it ok to get the following error while patching?:
```
error: patch failed: aten/src/THC/THCTensorRandom.cpp:83
error: aten/src/THC/THCTensorRandom.cpp: patch does not apply
```

ezyang commented on amd_build/hip_files/ATen/CMakeLists.txt.hip:
```diff
@@ -0,0 +1,454 @@
+CMAKE_MINIMUM_REQUIRED(VERSION 2.8)
```
Is this something like, the original `CMakeLists.txt` was copied and then edited until it worked with HIP? What's our ongoing maintenance strategy for noting changes to CMakeLists.txt and getting them to HIP? Would it make sense to try to reintegrate these into the real CMakeLists.txt and having search for HIP be controlled by a flag?

Jorghi12 replied:
The CMakeLists.txt files are extremely similar and they already contain a special WITH_ROCM flag (so does the patched setup.py file). They could both be integrated accordingly.

ezyang commented on amd_build/hip_files/THC/THCApply.cuh.hip:
```diff
@@ -0,0 +1,878 @@
+#ifndef THC_APPLY_INC
+#define THC_APPLY_INC
```
Same here: it would be good to know (via a comment at the top of the file), what the substantive differences of this file and `THCApply.cuh` are.

ezyang commented:
Looking at the patch, it seems sufficiently big that the next time someone merges a change to anything in THC, it will immediately stop working.

In general, I am wondering, would it make more sense to maintain this as a branch, until we can mainline the source code changes so that they work for AMD as well as CUDA? (I was not privy to any of the original discussions, so I apologize if this is something you've already looked at.)

Commit:
Updating CWD

Use the actual path for the file instead of the current working directory, which depends on where the script is invoked.

Jorghi12 committed:
Updating folder path for amd_build

Jorghi12 committed:
Removing previous amd_build directory

Jorghi12 committed:
Updated setup.py to support WITH_ROCM

Jorghi12 committed:
Renaming the files for CuDNN BatchNorm & Conv since having two .cpp files with the same name results in a linking error in the HCC compiler used for ROCm/AMD.

Jorghi12 committed:
Removing old BatchNorm & Conv files since they've been renamed.

Jorghi12 committed:
Updating build path to handle ROCM

Jorghi12 committed:
Cleaned up the build path and created a FindHIP cmake file for setting up relevant hip paths.

Jorghi12 committed:
Seperated the individual patch files to make it easier to detect issues while building.

Jorghi12 committed:
Removed CMakeLists hip files and fixed directory structure

Jorghi12 committed:
Merge branch 'sending_pr' of github.com:wsttiger/pytorch into sending_pr

Jorghi12 committed:
Adding build pytorch amd script

Jorghi12 committed:
Merged setup patch into PyTorch setup.py & cleaned a few issues

Jorghi12 committed:
Added information on where to download the hipify-python script.

Jorghi12 committed:
Resolved linting issues inside of build_pytorch_amd.py

Jorghi12 commented:
Integrated building for HIP into the aten/CMakeLists.txt, aten/src/ATen/CMakeLists.txt, setup.py, and tools/build_pytorch_libs.sh scripts, using the WITH_ROCM flag. And divided the giant patch file into separate individual patches per file in the codebase.

@ezyang - Plan is to merge into master and over time reduce the number of patch files required by making changes to the HIP code base or Hipify Script.

ezyang commented on aten/CMakeLists.txt:
```diff
@@ -337,6 +341,9 @@ int main()
 " HAS_C11_ATOMICS)
 
+# For ROCm stack, the hcc/hipcc compiler cannot use C11 atomics. Prefer GCC.
+SET(HAS_C11_ATOMICS OFF)
```
This SET() is being applied unconditionally, even if you're not using the ROCM stack.

ezyang commented on aten/CMakeLists.txt:
```diff
@@ -94,18 +98,18 @@ IF(MSVC)
 ENDIF(MSVC)
 
-IF (NOT MSVC)
+IF (NOT MSVC AND NOT WITH_ROCM)
```
I don't understand this conditional; does ROCM not support C++11?

ezyang commented on aten/CMakeLists.txt:
```diff
@@ -297,11 +301,11 @@ ENDIF(C_SSE2_FOUND)
   SET(CMAKE_C_FLAGS "${C_SSE4_1_FLAGS} -DUSE_SSE4_1 ${C_SSE4_2_FLAGS} -DUSE_SSE4_2 ${CMAKE_C_FLAGS}")
 ENDIF()
-IF(C_SSE3_FOUND)
+IF(C_SSE3_FOUND AND NOT WITH_ROCM)
```
Also confused about this conditional. Does ROCM not support SSE3?

ezyang commented on aten/CMakeLists.txt:
```diff
@@ -455,17 +462,22 @@ include_directories(
-  SET(CUDA_FLAG -n)
-  SET(AT_CUDA_ENABLED 0)
-else()
+if (WITH_ROCM)
```
I really want to see a message here saying that ROCM is eanbled

ezyang commented on aten/cmake/FindHIP.cmake:
```diff
@@ -0,0 +1,45 @@
+ADD_DEFINITIONS(-DNDEBUG)
+
+# Find the HIP Package
+FIND_PACKAGE(HIP 1.0 REQUIRED)
```
Generally, the cmake file which is actually invoked by `find_package` is given the name `Find*.cmake`. This file is named like this but doesn't follow that convention. It would be clearer to name it something like `LoadHIP.cmake`; just don't use the Find convention.

ezyang commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -1,7 +1,27 @@
+
+  # Set C++ Flags
+  SET(CMAKE_CXX_COMPILER ${HIP_HIPCC_EXECUTABLE})
+  SET(CMAKE_CXX_FLAGS  "-std=c++11 ${CMAKE_CXX_FLAGS} ${HIP_CXX_FLAGS}")
```
And now HIP seems to support C++11?

ezyang commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -246,9 +268,10 @@ set_property(TARGET tbb_static tbb_def_files PROPERTY FOLDER "dependencies")
 
-if(NOT ${CMAKE_VERSION} VERSION_LESS "3.1")
+# ROCM's HCC compiler fails if CXX_STANDARD==11 (HCC uses C++ AMP)
+if(NOT ${CMAKE_VERSION} VERSION_LESS "3.1" AND NOT WITH_ROCM)
```
And now it doesn't support C++11?

ezyang commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -354,6 +377,16 @@ IF(CUDA_FOUND)
 ENDIF()
 
+### Link in the ROCm stuff ###
+FIND_LIBRARY(HIPBLAS_LIBRARY hipblas HINTS ${HIPBLAS_PATH}/lib)
```
This looks like it is being unconditionally searched for. You didn't REQUIRE it, but wouldn't it be safer to only search if WITH_ROCM?

ezyang commented on setup.py:
```diff
@@ -753,6 +767,41 @@ def run(self):
+    hcc_include_path = '/opt/rocm/hcc/include'
+    hipblas_include_path = '/opt/rocm/hipblas/include'
+    hipsparse_include_path = '/opt/rocm/hcsparse/include'
+    print(rocm_include_path)
```
These look like debug prints

ezyang commented:
OK. @Jorghi12 can you modify `.travis.yml` to add another build configuration where it applies all of the ROCM diffs, and fails if they don't apply cleanly? You can skip this if we're supposed to be getting the AMD CI online in the very short future.

FYI, I think a decent number of patches are going to start tripping over AMD specific code and/or make these patches stop applying. We're going to need clear guidance about what to do in this situation (whether or not it's blocking the PR until you get around to it, or educating developers enough about the AMD stack so they can fix it themselves, or temporarily disabling functionality if it gets out of sync.)

apaszke commented on setup.py:
```diff
@@ -126,6 +126,16 @@
+    WITH_ROCM = True
+    WITH_CUDA = False
+
+    # If using ROCM stack disable distributed for now
```
Why? A hint in the comment would be useful

apaszke commented on setup.py:
```diff
@@ -822,6 +871,7 @@ def make_relative_rpath(path):
     DL = Extension("torch._dl",
                    sources=["torch/csrc/dl.c"],
                    language='c',
+                   extra_link_args=['-shared'] if WITH_ROCM else []
```
Why does `ROCM` affect this one? It's completely unrelated to any GPU stuff

Jorghi12 replied:
On the ROCm stack, the HCC compiler generates both the host & device code. I briefly spoke with a few developers at AMD about allowing other compilers such as GCC to be used on host - though it's not on their road map. It'll require modifications to [HIPCC](https://github.com/ROCm-Developer-Tools/HIP/blob/master/bin/hipcc). I'll give it a dive soon.

(HIPCC is what we set as the compiler in CMakeLists.txt)

apaszke commented on tools/amd_build/patches/a_aten_src_ATen_Error.h.patch:
```diff
@@ -0,0 +1,48 @@
+ #include <stdarg.h>
+ 
+-#if !defined(_WIN32)
++#if !defined(__HIP_PLATFORM_HCC__)
```
Looks like this will break the Windows build. Why can't you make it `!defined(_WIN32) && !defined(__HIP_PLATFORM_HCC__)`?

apaszke commented on tools/amd_build/patches/a_aten_src_ATen_cuda_CUDAHalf.cu.patch:
```diff
@@ -0,0 +1,19 @@
+ #include <cuda.h>
+ #include <cuda_runtime.h>
+ #include <cuda_fp16.h>
++#if !defined(__HIP_PLATFORM_HCC__)
```
If this file is completely disabled in the HIP build, then maybe it's just simpler to exclude it from the build?

apaszke commented on tools/amd_build/patches/a_aten_src_ATen_native_cuda_Distributions.cu.patch:
```diff
@@ -0,0 +1,36 @@
+   AT_DISPATCH_FLOATING_TYPES(ret.type(), "poisson", [&] {
+      poisson_cuda_kernel<scalar_t>(ret, lambda_, next_philox_seed(gen));
+    });
++  #endif
```
This looks really dangerous, because it will basically silently return an uninitialized tensor if you compile with HIP

apaszke commented on tools/amd_build/patches/a_aten_src_ATen_native_cuda_Embedding.cu.patch:
```diff
@@ -0,0 +1,27 @@
+       v += x * x;
+     } else {
+-      v += std::pow(x, norm_type);
++      v += powf(x, norm_type);
```
Those functions should be templated on the input type (like the whole kernel is), but have hardcoded `float`

apaszke commented on tools/amd_build/patches/a_aten_src_THCS_generic_THCSTensorMath.cu.patch:
```diff
@@ -0,0 +1,31 @@
+ 
+ void THCSTensor_(hspmm)(THCState *state, THCSTensor *r_, real alpha, THCSTensor *sparse_, THCTensor *dense) {
++#if defined(__HIP_PLATFORM_HCC__)
++  #define THRUST_EXEC(fn, ...) // whitespace
```
This also looks like it will silently start returning incorrect results. Would be better to replace `// whitespace` with `throw ...`

apaszke commented on tools/amd_build/patches/a_aten_src_THCUNN_BCECriterion.cu.patch:
```diff
@@ -0,0 +1,27 @@
++template <>
++inline __host__ __device__ float eps() { return 1e-12f; }
++template <>
++inline __host__ __device__ double eps() { return 1e-12; }
```
This code looks like it's what's originally been there, except for the `__host__` addition. We could just patch the original file if that helps.

apaszke commented on tools/amd_build/patches/a_aten_src_THCUNN_SoftMaxCommon.cuh.patch:
```diff
@@ -0,0 +1,18 @@
+   uint32_t block_threads = block.x * block.y;
+   smem_size = block.x == 1 ? 0 : block_threads * sizeof(AccumT);
++#if defined(__HIP_PLATFORM_HCC__)
++  int max_active_blocks = 8;
```
Maybe set this to something larger? 8 blocks is not much

apaszke commented on tools/amd_build/patches/a_aten_src_THC_THCAllocator.c.patch:
```diff
@@ -0,0 +1,13 @@
+   // on various compute capabilities.
+   void* ptr;
+-  THCudaCheck(cudaMallocManaged(&ptr, size, cudaMemAttachGlobal));
++  THCudaCheck(cudaSuccess);
```
This should also be replaced with `throw ...` instead of silently saying everything went file

apaszke commented on tools/amd_build/patches/a_torch_csrc_generic_StorageSharing.cpp.patch:
```diff
@@ -0,0 +1,28 @@
++  #undef PyBytes_GET_SIZE(op)
++  #define PyBytes_AS_STRING(op) (((PyBytesObject *)(op))->ob_sval)
++  #define PyBytes_GET_SIZE(op)  Py_SIZE(op)
++#endif
```
But why? Those shouldn't matter for HIP. This file should be still built using the regular C++ compiler

apaszke commented on tools/amd_build/patches/a_torch_csrc_utils_python_strings.h.patch:
```diff
@@ -0,0 +1,18 @@
++  #undef PyBytes_GET_SIZE(op)
++  #define PyBytes_AS_STRING(op) (((PyBytesObject *)(op))->ob_sval)
++  #define PyBytes_GET_SIZE(op)  Py_SIZE(op)
++#endif
```
Same here

apaszke reviewed (commented):
Nice work @Jorghi12! I do also have a few concerns about our current strategy, and the contents of the patches:
1. Why did we have to change so much of code that's completely unrelated to GPU? A few examples:
  - TH changes
  - The `docs/` directory
  - Many docstrings in Python files
2. There are some things that I don't quite understand:
  - Why do files in the `cudnn/` directory in ATen gained an `CuDNN` suffix? It's clear from the directory name and seems unnecessary. That's not a patch, so it will change how the repo looks like.
  - Why do we change Caffe2? We probably don't want to patch that.
3. Why do we need those `.hip` files. They look as if they were pretty much copied from THC, so why is it better to store them like this instead of patches? Some of them contain pretty complicated platform independent logic (think all the stride tricks in `THCApply`), so I'd rather avoid duplicating them.
4. I have very similar concerns to @ezyang. How long do we plan to keep the codebase in this state?

Jorghi12 committed:
Removing many unnecessary patch files. Removing unnecessary .hip files. Fixing up the build process.

ezyang commented:
@pytorchbot retest this please

Jorghi12 committed:
Refactored the PR for supporting HIP

apaszke commented:
Hmm what are the files that have the same file names and cause the conflicts? I can't see them.

Great stuff @Jorghi12 looking forward to more simplifications! Just FYI I think CUDA 9 also treats half as a pretty much first-class type, but we can't rely on that if we want to support CUDA 8.

wsttiger commented:
Hi @apaszke we're not modifying the original source, so the CUDA 8 (and below) half should be fine in THC.

wsttiger commented:
Having said that, perhaps this would be a good time add cross-platform support for __fp16 / half in THC.

apaszke commented:
@wsttiger oh yes, I know that. I only meant it as a general note. What do you mean by cross platform support for half?

Jorghi12 commented:
@apaszke Yeah you're right, the old BatchNorm.cpp isn't here anymore. The remaining duplicate is Conv.cpp which is [here](https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/mkldnn/Conv.cpp). To make things cleaner, I'll rename the Conv to ConvCuDNN using the preprocessor script (amd_build/build_pytorch_amd.py) and keep the original PyTorch file names.

Regarding the CUDA8/CUDA9, I see that makes sense.

ezyang commented:
@apaszke I still stand by my earlier suggestion. I don't see that there's anything wrong with naming the cudnn convolution CuDNNConv.cpp and the MKLDNN convolution MKLDNNConv.cpp. It seems like a reasonable enough bikeshed color to me, and if it simplifies the build I'm all for it.

apaszke commented:
@ezyang sorry, which suggestion do you have in mind?

Jorghi12 commented:
@pytorchbot retest this please

ezyang commented:
network flaked on OS X failure

wsttiger commented:
@apaszke I think that @Jorghi12 has implemented __half_raw which should be a cleaner implementation (then whatever we were doing before:))

Jorghi12 committed:
Minimizing the number of changes inside individual patches.

Commit:
Merge branch 'master' into sending_pr

Jorghi12 committed:
Cleaned up patch files.

Jorghi12 committed:
Merge branch 'sending_pr' of github.com:wsttiger/pytorch into sending_pr

Jorghi12 committed:
Removed patch files.

Jorghi12 committed:
Updating patches

Jorghi12 committed:
Removing HIP change from file.

Jorghi12 committed:
Cleaned up patches

ezyang commented on aten/CMakeLists.txt:
```diff
@@ -7,6 +7,10 @@ else()
 endif()
 
+IF(WITH_ROCM)
+  INCLUDE(${CMAKE_CURRENT_SOURCE_DIR}/cmake/LoadHIP.cmake)
```
Note for a follow up PR: we're moving all of the cmake modules into the top-level cmake directory, so LoadHIP should live there similarly.

ezyang commented on tools/amd_build/patches/a_aten_src_ATen_ATen.h.patch:
```diff
@@ -0,0 +1,9 @@
+--- a/aten/src/ATen/ATen.h
++++ b/aten/src/ATen/ATen.h
+@@ -1,3 +1,4 @@
++#include "hip/hip_runtime.h" 
```
This include runs counter to ATen header philosophy; we shouldn't include CUDA or HIP headers from the user visible ATen/ATen.h header.

ezyang commented on tools/amd_build/patches/a_aten_src_THC_THCTensorMathReduce.cuh.patch:
```diff
@@ -0,0 +1,69 @@
+ struct ReduceMin {
+   inline __device__ T operator()(T a, T b) const {
+-    return (THCNumerics<T>::lt(a, b) ||
++    return ((int)THCNumerics<T>::sub(a, b) < 0 ||
```
This int cast looks very questionable

Jorghi12 replied:
@wsttiger can explain further on this. There are issues with the compiler that we needed to work around.

ezyang commented on tools/amd_build/patches/a_aten_src_THCS_generic_THCSTensorMath.cu.patch:
```diff
@@ -0,0 +1,31 @@
+ accreal THCSTensor_(normall)(THCState *state, THCSTensor *self, real value) {
+   THCSTensor* self_coalesced = THCSTensor_(newCoalesce)(state, self);
+-  accreal result = THCTensor_(normall)(state, self_coalesced->values, value); 
++  accreal result = THCTensor_(normall)(state, self_coalesced->values, value);
```
No op?

ezyang commented on tools/amd_build/patches/a_aten_src_THCUNN_IndexLinear.cu.patch:
```diff
@@ -0,0 +1,64 @@
++#if defined(__HIP_PLATFORM_HCC__)
++__global__ 
++#else
+ __global__ static
```
Might be simplest to just unconditionally make all of these non-static.

ezyang replied:
Note about HIP limitations w.r.t. static kernels would be helpful, somewhere.

ezyang commented on tools/amd_build/patches/a_aten_src_THCUNN_generic_SpatialCrossMapLRN.cu.patch:
```diff
@@ -0,0 +1,19 @@
+   gradOutput = THCTensor_(newContiguous)(state, gradOutput);
+ 
+   int n_threads = batchSize * imsize_h * imsize_w;
++#if defined(__NVCC__)
```
So, HIP version of this just doesn't work, is that right? Wouldn't it be better to raise an error than just skip the kernel?

ezyang commented:
Thank you @Jorghi12, the new patches look much much improved! Just to summarize the things we're planning to fix later:

- Get AMD builders online. (Can we get more than one box from AMD, so that we're not perpetually low on capacity?)
- Eliminate more of the diffs
  - Don't `#include "hip/hip_runtime.h"` from user-facing headers
  - `&& !defined(__HIP_PLATFORM_HCC__)` tests don't have to be patched, we can unconditionally include them in our code
  - `WITH_ROCM` conditionals in cmake can be put straight in the cmake
  - We should probably make a `cuda_based_add_executable` which either switches to `cuda_add_executable` or `hip_add_executable`
  - Ifdef the bitlanes file
- Fix some correctness issues
  - Stop specializing abs/pow to float

Also, longer term thing I am thinking about: with https://github.com/pytorch/pytorch/pull/7275 coming down the pipe, it would be nice to think of a way where you can easily build CPU/CUDA/AMD all in one go into separate libraries, not controlled by a build flag.

Jorghi12 committed:
Added AVX/SSE avoidance due to bug with ROCms stack. Just temporary for now.

Jorghi12 committed:
Removing the other HIP file

Jorghi12 committed:
Removed patch file + merged ROCm into Aten/test

apaszke reviewed (commented):
Looks great!

Is THCTensorRandom.cu.hip really so much different than THCTensorRandom.cu? Why isn't it a patch.

Jorghi12 committed:
Removed ATen tests patch file and updated disbale_features yaml to remove headers that don't exist on the HIP stack.

Commit:
Merge branch 'master' into sending_pr

Jorghi12 committed:
Reduced the number of patches down to 14 after Edward's suggestions.

Jorghi12 committed:
Merge branch 'sending_pr' of github.com:wsttiger/pytorch into sending_pr

Jorghi12 committed:
Transferred deletion of certain functions from patch to yaml file.

wsttiger committed:
Set default Thrust path

Jorghi12 committed:
Fixed aten files so we now use the templated pow/abs instead of std:: directly.

Jorghi12 committed:
Merge branch 'sending_pr' of github.com:wsttiger/pytorch into sending_pr

Jorghi12 committed:
Removed error from aten/src/THCUNN/Abs.cu

Jorghi12 committed:
Updated the locations of the cmake build files. Moved THCTensorRandom from a hip to a patch file. Added executable/library commands that can successfully handle either CUDA or HIP.

Jorghi12 committed:
Removed hip extraction from the build script and removed the old hip file.

Jorghi12 committed:
Replaced MACRO with function in upper level cmake.

Jorghi12 committed:
Added empty ELSE() block to prevent the loading of a command without CUDA or HIP. Also added IF guards around torch_cuda_based_add_executable in Aten tests.

Jorghi12 committed:
Updated aten tests.

Jorghi12 committed:
Removed the hip include from the ATen header.

Jorghi12 commented:
Updates:
@apaszke I removed the hip files completely and left THCTensorRandom.cu in a minimal patch form. (Static casts are added due to the annoying implementation of hipKernelLaunch.)

Top level CMake Changes:
- The LoadHIP.cmake script has been added to cmake/public since it'll be useful for Caffe2 build as well.
- The command torch_cuda_based_add_executable was added inside of the cmake/public/utils.cmake. Similarly done for torch_cuda_based_add_library.

Jorghi12 committed:
Can't throw exceptions on C++ AMP, using abort

Jorghi12 committed:
Missing IF guards for cuda/hip executables in aten tests.

Jorghi12 committed:
Removed a series of patch files.

Jorghi12 committed:
Added template keyword to help out the HCC compiler.

Jorghi12 committed:
Rebased the specific files displayed in the PR

Commit:
Merge branch 'master' into sending_pr

Jorghi12 committed:
Fixing typo.

Jorghi12 committed:
Merge branch 'sending_pr' of github.com:wsttiger/pytorch into sending_pr

Jorghi12 committed:
Change flag from "WITH_CUDA" to "NOT NO_CUDA"

Replacing "WITH_CUDA" with "NOT NO_CUDA" after the rebase.

Jorghi12 committed:
Fix LoadHIP path

Jorghi12 committed:
Updating build files after rebasing.

Jorghi12 committed:
Reorganization after cpu/gpu separation.

Jorghi12 committed:
Removed HIPCC from setup.py & removed -shared extra linking args.

Commit:
Merge branch 'master' into sending_pr

Jorghi12 committed:
Updated CMake / Setup build to correctly link when under ROCm stack.

Jorghi12 committed:
Merge branch 'sending_pr' of github.com:wsttiger/pytorch into sending_pr

apaszke commented on setup.py:
```diff
@@ -822,6 +862,7 @@ def make_relative_rpath(path):
     DL = Extension("torch._dl",
                    sources=["torch/csrc/dl.c"],
                    language='c',
+                   extra_link_args=[]
```
Can you please remove this line?

apaszke reviewed (approved):
LGTM. One minor nit. Should be ready to merge once @ezyang approves the CMake part.

Jorghi12 committed:
Removed the unnecessary argument from Extension constructor.

yf225 commented:
@pytorchbot retest this please

Jorghi12 committed:
Adding another test to be included with ROCm building.

Jorghi12 committed:
Updated the setup_helpers scripts in order to get around linter error

Jorghi12 committed:
Fix syntax issue

Jorghi12 committed:
Solving lint issue: line too long

Jorghi12 commented:
@pytorchbot retest this please

Jorghi12 commented:
@pytorchbot retest this please

Jorghi12 commented:
The build for pr/pytorch-linux-xenial-py3-clang5-asan was hanging for 3 hours previously so I restarted the build on Jenkins: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-py3-clang5-asan-build/4652/console
